app/Services/BannerFinder.php

```diff
@@ -89,42 +85,57 @@ public static function getBestBanners(array $zones, array $keywords)
             foreach ($zones as $zoneInfo) {
                 $zone = Zone::find($zoneInfo['zone']);
 
-                try {
-                    // $zone instanceof Zone; // ?? Yodahack : what the hack
-                    $bannerIds[] = NetworkBanner::where('creative_width', $zone->width)
-                      ->where('creative_height', $zone->height)
-                      ->whereIn('creative_type', $typeDefault)
-                      ->get()->pluck('uuid')->random();
-                } catch (\InvalidArgumentException $e) {
-                    $bannerIds[] = '';
+                if ($zone) {
+                    try {
+                        // $zone instanceof Zone; // ?? Yodahack : what the hack
+                        $bannerIds[] = NetworkBanner::where('creative_width', $zone->width)
+                            ->where('creative_height', $zone->height)
+                            ->whereIn('creative_type', $typeDefault)
+                            ->get()->pluck('uuid')->random();
+                    } catch (\InvalidArgumentException $e) {
+                        $bannerIds[] = '';
+                    }
+                } else {
+                    $bannerIds[] = md5(rand());
                 }
             }
         }
 
         $banners = [];
         foreach ($bannerIds as $bannerId) {
-            // TODO:
-            $banner = $bannerId ? NetworkBanner::where('uuid', hex2bin($bannerId))->first() : null;
-
-            if (!empty($banner)) {
-                $campaign = NetworkCampaign::find($banner->network_campaign_id);
-                $banners[] = [
-                    'serve_url' => $banner->serve_url,
-                    'creative_sha1' => $banner->creative_sha1,
-                    'pay_from' => $campaign->adshares_address, // send this info to log
-                    'click_url' => route('log-network-click', [
-                        'id' => $banner->uuid,
-                        'r' => Utils::urlSafeBase64Encode($banner->click_url),
-                    ]),
-                    'view_url' => route('log-network-view', [
-                        'id' => $banner->uuid,
-                        'r' => Utils::urlSafeBase64Encode($banner->view_url),
-                    ]),
-                ];
-            } else {
-                $banners[] = null;
-                // TODO: discuss Jacek
-            }
+            $banners[] = [
+                'serve_url' => config('app.url') . '/img/logo_x.png',
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "app.url" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)